### DISCUSSION TITLE
Overload with booleans. Again...

### DISCUSSION THREAD

**Discussion by lorenzwalthert** [microsoft/pyright#6819] [Q&A] [answered]
My goal is to create type annotation for overloads. I took the [solution](https://github.com/microsoft/pyright/issues/1616#issuecomment-796320440) from #1616 (which is based on #428) and managed to get a this to pass pyright as expected:
```python
from typing import Literal, overload, Union
import typing

@overload
def myfunc(arg: Literal[True]) -> str:
    ...

@overload
def myfunc(arg: Literal[False]) -> int:
    ...

def myfunc(arg: bool) -> Union[int, str]:
    if arg:
        return "something"
    else:
        return 0

myfunc(True)
```
However, when I call the function `myfunc()` inside another function, I can't get rid of type errors. I add to the above file:
```python 
def f1(x: typing.Literal[True, False]):
    myfunc(x)

def f2(x: bool):
    myfunc(x)
```
And I get a type errors for both versions:
>   /path/to/file.py:21:5 - error: No overloads for "myfunc" match the provided arguments (reportGeneralTypeIssues)
  /path/to/file.py:21:12 - error: Argument of type "bool" cannot be assigned to parameter "arg" of type "Literal[False]" in function "myfunc"
    "bool" cannot be assigned to type "Literal[False]" (reportGeneralTypeIssues)
  /path/to/file.py:25:5 - error: No overloads for "myfunc" match the provided arguments (reportGeneralTypeIssues)
  /path/to/file.py25:12 - error: Argument of type "bool" cannot be assigned to parameter "arg" of type "Literal[False]" in function "myfunc"
    "bool" cannot be assigned to type "Literal[False]" (reportGeneralTypeIssues)

Can someone please guide me as to how to resolve the error? Thanks. I am using the latest release of pyright:
```bash
$ pyright --version
pyright 1.1.343
```

**Comment by erictraut** [comment, marked as answer]
For a variety of reasons, pyright collapses `Literal[True, False]` into `bool`. This is allowed because the two types are completely equivalent. That's why `f1` and `f2` in your example above behave the same.

When resolving overloads, pyright (like mypy) use a technique called "union expansion". For details, read [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=overloads). However, union expansion is not applied to non-union types like `bool` even if it can be expanded into a union of literals. Doing so would be costly at analysis time, so I'm reluctant to do it.

Note also that the technique used for overload matching is under-specified in the Python typing spec, and every type checker does it a little differently. There is an effort underway to expand the typing spec to remove ambiguities like this. Once the overload specification is complete, it's likely that all type checkers (including pyright) will need to make small changes to conform to the standard. If union expansion for `bool` is part of the final spec, I will implement it in pyright. If it's not, then I won't (since doing so would be non-compliant).

As for your particular case, my recommendation is to add a fallback overload that accepts an argument of type `bool`.

```python
@overload
def myfunc(arg: bool) -> int | str:
    ...
```

**Reply by lorenzwalthert** [reply]
Thanks @erictraut for your fast and detailed answer. I have skimmed over a few other issues here in the past and I wanted to tell you that your answers are all highly valuable and I think everyone coming here appreciates your issue management. 6 open issues for a repo that big is also exceptional. Keep it up 👍
